runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/TreeJsonInput.kt

```diff
@@ -116,16 +119,16 @@ private open class JsonTreeInput(json: Json, override val obj: JsonObject) : Abs
     override fun currentElement(tag: String): JsonElement = obj.getValue(tag)
 
     override fun endStructure(desc: SerialDescriptor) {
-        // This one can be optimized
-        if (json.strictMode) {
-            val names = HashSet<String>(desc.elementsCount)
-            for (i in 0 until desc.elementsCount) {
-                names += desc.getElementName(i)
-            }
+        if (!json.strictMode || desc is PolymorphicClassDescriptor) return
```

Comment:
I have a feeling that `desc is PolymorphicClassDescriptor` would break this validation when deserializing from JsonElement to an arbitrary data class.